Prevent stacking of power requests

daniel-zullo-frequenz opened the pull request:
The power distributing actor processes one power request at a time to prevent multiple requests for the same components from being sent to the microgrid API concurrently. Previously, this could lead to the request channel receiver becoming full if the power request frequency was higher than the processing time. Even worse, the requests could be processed late, causing unexpected behavior for applications setting power requests. Moreover, the actor was blocking power requests with different sets of components from being processed if there was any existing request.
    
This patch ensures that the actor processes one request at a time for different sets of components and keeps track of the latest pending request if there is an existing request with the same set of components being processed. The pending request will be overwritten by the latest received request with the same set of components, and the actor will process it once the request with the same components is done processing.

daniel-zullo-frequenz committed:
Prevent stacking of power requests

The power distributing actor processes one power request at a time
to prevent multiple requests for the same components from being sent
to the microgrid API concurrently. Previously, this could lead to
the request channel receiver becoming full if the power request
frequency was higher than the processing time. Even worse, the
requests could be processed late, causing unexpected behavior
for applications setting power requests. Moreover, the actor was
blocking power requests with different sets of components from being
processed if there was any existing request.

This patch ensures that the actor processes one request at a time
for different sets of components and keeps track of the latest pending
request if there is an existing request with the same set of
components being processed. The pending request will be overwritten
by the latest received request with the same set of components,
and the actor will process it once the request with the same components
is done processing.

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

daniel-zullo-frequenz committed:
Update power distributor documentation

The documentation was updated to reflect the current
state of the actor.

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

daniel-zullo-frequenz committed:
Update release-notes

To add an entry to release-nots related to preventing stacking
of power requests to avoid delays in processing when the request
frequency exceeds the processing time.

Signed-off-by: Daniel Zullo <daniel.zullo@frequenz.com>

daniel-zullo-frequenz commented:
@shsms  or @llucax would you mind to have an initial look at this draft?

shsms commented on src/frequenz/sdk/actor/power_distributing/power_distributing.py:
```diff
@@ -132,8 +132,32 @@ async def _run(self) -> None:  # pylint: disable=too-many-locals
+
         async for request in self._requests_receiver:
-            await self._component_manager.distribute_power(request)
+            task_name = f"power_request_ids={request.component_ids}"
```
`request.component_ids` is a set, don't think it would be stable, so we might end up sending overlapping requests.

Converting to a frozenset and having a frozenset as key sounds more efficient that converting to list, sorting, converting to string.

daniel-zullo-frequenz replied:
I updated to use a frozenset but I still need the task name for lookups once a task processed/done

shsms replied:
I think if you need a string, you need to convert to list, then sort, then convert to string.  Else the generated strings will have component ids in unpredictable order.

daniel-zullo-frequenz replied:
What do you think about computing the hash of a frozenset as they are hasheable? The main disadvantages I can think of are it won't be helpful information for debugging and it isn't reproducible in different platforms/environments. But it should be cheaper the conversion plus sorting and it would be a relatively short string

daniel-zullo-frequenz replied:
I've just sorted it at the moment which converts it to a sorted list

shsms replied:
okay, using the hash sounds like a good idea as well.  But I'll let you decide.

llucax commented on src/frequenz/sdk/actor/power_distributing/power_distributing.py:
```diff
@@ -132,8 +132,32 @@ async def _run(self) -> None:  # pylint: disable=too-many-locals
+
+        def cleanup(task: asyncio.Task[None]) -> None:
+            task_name = task.get_name()
+            assert task_name in power_tasks, "Task not found in power_tasks"
```
I would probably log an error here and return, I guess bringing the whole thing down doesn't make sense only because we can't clean something up. Or maybe this is just ignored, because is called via `add_done_callback()` :thinking:

In any case logging an error I think is the right approach.

daniel-zullo-frequenz replied:
I was mainly using this to find programming errors, the logic is wrong if this ever happens. Anyway we can also log it as an error

daniel-zullo-frequenz replied:
Updated

llucax replied:
Yeah, I agree the logic is wrong, but it doesn't seem like a very bad thing if it happens for some reason, it only means someone, somehow, removed the task, or it wasn't added in the first place, but we can keep working with what we have.

daniel-zullo-frequenz replied:
Right, I previously saw your point and I updated it. The assertion was mainly useful for me to code and debug the patch-fix

llucax commented on src/frequenz/sdk/actor/power_distributing/power_distributing.py:
```diff
@@ -132,8 +132,32 @@ async def _run(self) -> None:  # pylint: disable=too-many-locals
+            assert task_name in power_tasks, "Task not found in power_tasks"
+
+            # Raise if there was an exception while processing the task
+            task.result()
```
Same here, just log the exception, and continue with the cleanup (remove the task from the dict).

daniel-zullo-frequenz replied:
Fixed

llucax commented on src/frequenz/sdk/actor/power_distributing/power_distributing.py:
```diff
@@ -132,8 +132,32 @@ async def _run(self) -> None:  # pylint: disable=too-many-locals
+
         async for request in self._requests_receiver:
-            await self._component_manager.distribute_power(request)
+            task_name = f"power_request_ids={request.component_ids}"
```
Maybe add more info, like a prefix with `{type(self).__name__}:`.

daniel-zullo-frequenz replied:
makes sense. I was also considering to hash the list component IDs as the list could be a long list of components

daniel-zullo-frequenz replied:
Updated as suggested

llucax commented on src/frequenz/sdk/actor/power_distributing/power_distributing.py:
```diff
@@ -132,8 +132,32 @@ async def _run(self) -> None:  # pylint: disable=too-many-locals
+                self._component_manager.distribute_power(request),
+                name=task_name,
+            )
+            power_tasks[task_name].add_done_callback(cleanup)
```
This seems to actually be a perfect use case for a class I'm planning to add to `core` soon :tm: as part of the revamp of background service (I'm splitting out the task handling code to have a *TaskGroup* that is more useful for us for cases like this).

daniel-zullo-frequenz replied:
Sounds great!

llucax replied:
* https://github.com/frequenz-floss/frequenz-core-python/pull/30

llucax commented on src/frequenz/sdk/actor/power_distributing/power_distributing.py:
```diff
@@ -132,8 +132,32 @@ async def _run(self) -> None:  # pylint: disable=too-many-locals
         async for request in self._requests_receiver:
-            await self._component_manager.distribute_power(request)
+            task_name = f"power_request_ids={request.component_ids}"
+            if task_name in power_tasks:
+                result = Error(
+                    request=request,
+                    msg="A request for the same components is being processed. Skipped.",
+                )
+                await self._result_sender.send(result)
+                continue
```
Wouldn't it be better to cancel the ongoing request and do the new one instead? Maybe not because we can end up with no request being finished if we have a burst of requests?

shsms replied:
I agree that the latest one needs to be executed, but instead of cancelling the old one, I think we should keep the latest request and send it as soon as the running one is complete.  That is, discard only the older requests and execute the latest.

daniel-zullo-frequenz replied:
> Maybe not because we can end up with no request being finished if we have a burst of requests?

Indeed, my thought here was to avoid entering a loop of cancelling the in-progress request when a new request arrives. This is a bit tricky because we don't really know the state of the current processing request so I decided to ignore/skip the new request. This is the point we all need to agree on.  @shsms  any thoughts on this?

> Wouldn't it be better to cancel the ongoing request and do the new one instead?

According to the outdated documentation, the actor was doing this. And the current state is the actor is just queuing up requests (without this patch)

daniel-zullo-frequenz replied:
@shsms I haven't seen your comment when I replied to this thread

shsms replied:
Absolutely, I agree with the reasoning above on why cancelling doesn't make sense.  We shouldn't cancel.

daniel-zullo-frequenz replied:
Updated to keep track of the latest pending power request

llucax commented on tests/timeseries/_battery_pool/test_battery_pool_control_methods.py:
```diff
@@ -213,6 +213,8 @@ async def test_case_1(
             await bounds_rx.receive(), power=1000.0, lower=-4000.0, upper=4000.0
         )
 
+        await asyncio.sleep(0.0)
```
Please document why all these new sleeps are needed.

daniel-zullo-frequenz replied:
This was just a failed attempt to fix tests. I prefer to validate the code-changes rather than jumping in fixing tests and docs. This is the reason I keep it as a draft, to show and to discuss the changes. Sorry, I should have added a note/comment for this

daniel-zullo-frequenz replied:
Updated

llucax reviewed (commented):
I didn ´t check the docs updates yet, focusing on the code first.

shsms commented on src/frequenz/sdk/actor/power_distributing/power_distributing.py:
```diff
@@ -132,8 +132,32 @@ async def _run(self) -> None:  # pylint: disable=too-many-locals
-            await self._component_manager.distribute_power(request)
+            task_name = f"power_request_ids={request.component_ids}"
+            if task_name in power_tasks:
+                result = Error(
+                    request=request,
+                    msg="A request for the same components is being processed. Skipped.",
+                )
+                await self._result_sender.send(result)
```
I think we shouldn't call this an `Error`.  I guess we need a new result type.  Maybe `Discarded`.

But I would rather not send a message at all,  and keep the latest request and discard just the older ones, and execute the latest request as soon as the running one is complete.

daniel-zullo-frequenz replied:
> But I would rather not send a message at all, and keep the latest request and discard just the older ones, and execute the latest request as soon as the running one is complete.

Interesting. So you'd like to keep the current in-progress request (without cancelling it) and queue only the latest request that couldn't have been processed yet because there is a similar one in progress. It makes sense to me @llucax what do you think about this idea?

daniel-zullo-frequenz replied:
This is not longer sending the Error, instead it is keeping track of the latest received power request

llucax replied:
The idea of finishing the current request and keep only the latest one to execute next is great!

About not sending responses, I don't remember how responses work, but shouldn't you send a response for every request? If not then great, if yes I also agree `Discarded` is better than `Error`.

daniel-zullo-frequenz replied:
@shsms your input is needed here?

My understanding is it's only needed to send a response if the request is processed (either succeeded, partial failed, error, etc). The current patch works as a request replacement if the previous request was queued up

shsms replied:
Yes.  Ever since the power manager was introduced, there has not been a one-to-one mapping between user proposals and response from the power distributor.  So just a log message should be enough saying new request received before previous one could start executing, so it is discarded.  And I think we have something of this form.

shsms commented on src/frequenz/sdk/actor/power_distributing/power_distributing.py:
```diff
@@ -1,14 +1,12 @@
+The purpose of this actor is to distribute power between components in a microgrid.
 
-Purpose of this actor is to keep SoC level of each component at the equal level.
+The actor receives power requests from component managers, process them distributing
```
Wow, this has been outdated for a long time.

```suggestion
The actor receives power requests from the power manager, process them by distributing

shsms commented on src/frequenz/sdk/actor/power_distributing/power_distributing.py:
```diff
@@ -32,28 +30,23 @@
-    The purpose of this tool is to keep an equal SoC level in all batteries.
-    The PowerDistributingActor can have many concurrent users which at this time
-    need to be known at construction time.
+    The actor can have many concurrent component managers.
+    For each component manager a bidirectional channel needs to be created
+    through which they can send and receive requests and responses.
```
I think this needs some changes.  One instance of the PowerDistributingActor can handle one component category.  The component cagegory and type need to be specified at actor startup and it will setup the correct component manager based on the given cagegory.

There are no bidirectional channels anymore.  The PowerDistributingActor takes a receiver for receiving requests and a sender for sending results.

daniel-zullo-frequenz replied:
Updated based on the suggestion

shsms commented on src/frequenz/sdk/actor/power_distributing/power_distributing.py:
```diff
@@ -32,28 +30,23 @@
-    It is recommended to wait for PowerDistributingActor output with timeout. Otherwise if
-    the processing function fails then the response will never come.
+    It is recommended to wait for PowerDistributingActor output with timeout.
+    Otherwise if the processing function fails then the response will never come.
```
I think these lines can go away, this was a quirk of bidirectional channels.  Now the request and results are decoupled, and results are directly available to users, but users don't send direct requests anymore.

daniel-zullo-frequenz replied:
gone

shsms commented on src/frequenz/sdk/actor/power_distributing/power_distributing.py:
```diff
@@ -32,28 +30,23 @@
+    * A new power request will be ignored if a power request with the same
+    components is currently being processed.
+
+    * If there are 2 requests and their components is different but they overlap
+    (they have at least one common component), then both requests will be
+    processed.
```
The power manager will make sure this doesn't happen, but I think we should ensure there are no overlaps, else requests should be queued up (just the latest, discarding older queue items).

But creating an issue and handling later is fine, because overlapping component IDs are not possible at the moment, with the power manager.

daniel-zullo-frequenz replied:
https://github.com/frequenz-floss/frequenz-sdk-python/issues/1030

daniel-zullo-frequenz commented:
@llucax @shsms I haven't finished updated the documentation. Would you mind to have another look just focusing on the code?

daniel-zullo-frequenz commented:
Updated documentation

llucax commented on src/frequenz/sdk/actor/power_distributing/power_distributing.py:
```diff
@@ -121,19 +133,37 @@ def __init__(  # pylint: disable=too-many-arguments
+                    )
+                self._pending_requests[name] = request
+            else:
+                self._process_request(name, request)
```
Super tiny but you can avoid a double-lookup if you use this pattern:

```py
if pending_request := self._pending_requests.get(name):
_logger.warning(
    "Pending request: %s, overwritten with request: %s", pending_request, request
)
```

daniel-zullo-frequenz replied:
Updated as suggested

llucax commented:
Just as a side note, I find the code a bit convoluted, but I think we can address that in the future.

daniel-zullo-frequenz commented:
Replaced the task name previously used as a key to identify processing and pending requests. The request ID, which is a frozenset of the component IDs, is now used instead. This should make the code flexible to address #1030

daniel-zullo-frequenz commented:
> Just as a side note, I find the code a bit convoluted, but I think we can address that in the future.

Ok, let me know if you have some ideas to improve it and I can create an issue at least.
I failed my attempt to use the asyncio high level functionality given our use-case. I also tried to use the tasks set provided by the actor but unfortunately it wasn't suitable for our use-case.

llucax commented:
> Ok, let me know if you have some ideas to improve it and I can create an issue at least.

I actually tried to suggest something where we abstract the *ongoing requests* in an object that holds the task and the pending request (if any), but it was still complicated to implement like that because how we need to keep track of the tasks.  I think the new `PersistentTaskGroup` in `core` can help a lot here, but it is not there yet :)

daniel-zullo-frequenz commented:
> I actually tried to suggest something where we abstract the ongoing requests in an object that holds the task and the pending request (if any), but it was still complicated to implement like that because how we need to keep track of the tasks. I think the new PersistentTaskGroup in core can help a lot here, but it is not there yet :)

I had a quick look and make a comment [there](https://github.com/frequenz-floss/frequenz-core-python/pull/30#issuecomment-2277469054) to check if we can make it 100% suitable for the use case in this PR. Please be aware that I probably misunderstood the `PersistentTaskGroup` capabilities.

In any case I can create an issue in the SDK to refactor the PowerDistributingActor using  `PersistentTaskGroup`.

llucax commented:
I actually stayed on this issue thinking about it again...

[.... one hour later due to some of the changes you did with the request ID :D ...]

It was a good exercise to see how to apply the new class. The result is pretty different to what I had in mind before, and it doesn't look particularly simple either in retrospect, but I think it is more efficient, because we only keep one task for the same `req_id`, as long as we have quests queued up, instead of starting and finishing a task for every pending request.

```py
    async def _run(self) -> None:
        await self._component_manager.start()

        async with PersistentTaskGroup() as group:
            async for request in self._requests_receiver:
                self._handle_request(request, group)
                async for task in group.as_completed(timeout=0):
                    try:
                        task.result()
                    except Exception:  # pylint: disable=broad-except
                        _logger.exception("Failed power request: %s", request)

    async def _handle_request(self, request: ..., group: ...) -> None:
        req_id = self._get_request_id(request)

        if pending_request := self._pending_requests.get(req_id):
            _logger.debug(
                "Pending request: %s, overwritten with request: %s",
                pending_request,
                request,
            )
            return

        self._pending_requests[req_id] = request
        task = group.create_task(
            self._distribute_power(req_ir),
            name=f"{type(self).__name__}:{request}",
        )

    async def _distribute_power(self, req_id: ...) -> None:
        while request := self._pending_requests.get(req_id):
            self._component_manager.distribute_power(request)
```

At some point I want to add a way to get `as_completed()` as a `Receiver`, so we can put it in a `select()`, like:

```py
        async with PersistentTaskGroup() as group:
            completed_tasks = CompletedTaskReceiver(group)  # This needs to be added to channels
            async for selected in select(self._requests_receiver, completed_tasks):
                if selected_from(self._requests_receiver, selected):
                    self._handle_request(selected.message, group)
                elif selected_from(completed_tasks, selected):
                    try:
                        task.result()
                    except Exception:  # pylint: disable=broad-except
                        _logger.exception("Failed power request: %s", request)
```

But for now we can live with using it with `timeout=0` so it doesn't block, and we only ACK finished tasks after we handle a power request, which should be OK, we should receive requests often enough I guess.

llucax reviewed (commented):
LGTM. Will leave the final approval to @shsms as he had some comments too.

daniel-zullo-frequenz commented:
@llucax the final result looks great! I'll create an issue referencing your draft to address it once `PersistentTaskGroup` is available

llucax commented on src/frequenz/sdk/actor/power_distributing/power_distributing.py:
```diff
@@ -121,19 +133,34 @@ def __init__(  # pylint: disable=too-many-arguments
+
+            if req_id in self._processing_tasks:
+                if pending_request := self._pending_requests.get(req_id):
+                    _logger.warning(
```
A small one:

```suggestion
                    _logger.debug(
```

Otherwise I think it could get pretty spammy. There are many reasons why a request could get ignored, so I don't think we need to emit a warning for this, it is normal.

daniel-zullo-frequenz replied:
Makes sense. Thanks!

daniel-zullo-frequenz committed:
Update log type for ignored requests

Changed log type for ignored requests to avoid spamming, as there are many reasons a request
might be ignored.

Co-authored-by: Leandro Lucarella <luca@llucax.com>
Signed-off-by: daniel-zullo-frequenz <120166726+daniel-zullo-frequenz@users.noreply.github.com>

shsms reviewed (dismissed):
LGTM as well.

daniel-zullo-frequenz commented:
Thanks, will rebase it onto latest v1.x.x since there are some conflicts with the release notes

daniel-zullo-frequenz commented:
Just for the record, created https://github.com/frequenz-floss/frequenz-sdk-python/issues/1032 to improve the actor code
